prymer/offtarget/offtarget_detector.py

```diff
@@ -334,27 +353,74 @@ def _build_off_target_result(
-        p1: BwaResult = hits_by_primer[primer_pair.left_primer.bases]
-        p2: BwaResult = hits_by_primer[primer_pair.right_primer.bases]
-
-        # Get all possible amplicons from the left_primer_mappings and right_primer_mappings
-        # primer hits, filtering if there are too many for either
-        if p1.hit_count > self._max_primer_hits or p2.hit_count > self._max_primer_hits:
+        left_bwa_result: BwaResult = hits_by_primer[primer_pair.left_primer.bases]
+        right_bwa_result: BwaResult = hits_by_primer[primer_pair.right_primer.bases]
+
+        # If there are too many hits, this primer pair will not pass. Exit early.
+        if (
+            left_bwa_result.hit_count > self._max_primer_hits
+            or right_bwa_result.hit_count > self._max_primer_hits
+        ):
             result = OffTargetResult(primer_pair=primer_pair, passes=False)
-        else:
-            amplicons = self._to_amplicons(p1.hits, p2.hits, self._max_amplicon_size)
-            result = OffTargetResult(
-                primer_pair=primer_pair,
-                passes=self._min_primer_pair_hits <= len(amplicons) <= self._max_primer_pair_hits,
-                spans=amplicons if self._keep_spans else [],
-                left_primer_spans=(
-                    [self._hit_to_span(h) for h in p1.hits] if self._keep_primer_spans else []
-                ),
-                right_primer_spans=(
-                    [self._hit_to_span(h) for h in p2.hits] if self._keep_primer_spans else []
-                ),
+            if self._cache_results:
+                self._primer_pair_cache[primer_pair] = replace(result, cached=True)
+            return result
+
+        # Get the set of reference names with hits
+        hits_by_refname: dict[str, PrimerPairBwaHitsBySideAndStrand] = {
+            hit.refname: PrimerPairBwaHitsBySideAndStrand()
+            for hit in left_bwa_result.hits + right_bwa_result.hits
+        }
+
+        # Split the hits for left and right by reference name and strand
+        for hit in left_bwa_result.hits:
+            if hit.negative:
+                hits_by_refname[hit.refname].left_negative.append(hit)
+            else:
+                hits_by_refname[hit.refname].left_positive.append(hit)
+
+        for hit in right_bwa_result.hits:
+            if hit.negative:
+                hits_by_refname[hit.refname].right_negative.append(hit)
+            else:
+                hits_by_refname[hit.refname].right_positive.append(hit)
+
+        # Build amplicons from hits on the same reference with valid relative orientation
+        amplicons: list[Span] = []
+        for hits in hits_by_refname.values():
+            amplicons.extend(
+                self._to_amplicons(
+                    positive_hits=hits.left_positive,
+                    negative_hits=hits.right_negative,
+                    max_len=self._max_amplicon_size,
+                    strand=Strand.POSITIVE,
```

Comment:
I'm not sure how to think about strand here.  I think it's probably mostly irrelevant and could just always be set to positive?

Reply:
In the `PrimerPair` code, it's not explicitly enforced by a `__post_init__` but is strongly assumed by e.g. [calculate_amplicon_span](https://github.com/fulcrumgenomics/prymer/blob/2656adf87e814c80111f7b6272ce38787119a4dc/prymer/api/primer_pair.py#L223C9-L223C32) that the left primer is left of the right primer. This implies that the left primer is on the positive strand and the right primer is on the negative strand. All amplicons are coded as on the positive strand (default value for `Span`, again, not explicit).

When a hit to a primer pair is in the same orientation as the primer pair definition, e.g. left positive and right negative, then it makes sense for the amplicon to also be on the positive strand.

If the hit is in the opposite orientation to the primer pair definition, the amplicon should be on the negative strand.

I think it makes sense to assign strandedness like this:

- left primer + / right primer - = amplicon +
- left primer + / left primer - = amplicon +
- right primer + / left primer - = amplicon -
- right primer + / right primer - = amplicon -

Reply:
As discussed elsewhere, will leave questions of strand and L/L R/R hits to a separate PR.